### ISSUE TITLE
Literal instance for Float types is not downward-closed

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/cryptol#1062]
The Cryptol typechecker assumes that `Literal` class instances are downward-closed (i.e. that `Literal m a` implies `Literal n a` for `n < m`) but this is not true for the `Float` types. This leads to weird situations like this:
```
Cryptol> :m Float
Loading module Cryptol
Loading module Float
Float> let xs = [6,7,8]
Float> :t xs
xs : {a} (Literal 8 a) => [3]a
Float> [6,7,8] : [3](Float 3 2)

[error] at <interactive>:4:4--4:5:
  • `7` is not a valid literal of type `Float 3 2`
      arising from
      use of literal or demoted expression
      at <interactive>:4:4--4:5
Float> xs : [3](Float 3 2)
[0x6.0, 0x8.0, 0x8.0]
```
So the expression `[6,7,8]` gives a type error at type `Float 3 2` because `Float 3 2` cannot represent 7, but it typechecks when we move `[6,7,8]` into a separate definition, and the 7 gets rounded up to 8.

**Comment by LeventErkok** [comment]
For a sequence like this, really not much you can do but convert each element in the list during _type-checking time_ and make sure the conversion was exact, and otherwise reject the sequence I guess. That can be costly for longer sequences, but hopefully those don't come up very often.

**Comment by brianhuffman** [comment]
At type-checking time for `xs`, we can't check that the conversion is exact, because it's a polymorphic value, and we don't know what the eventual element type might be.

The question to answer is, what should be the type of a polymorphic value like `[6,7,8]`? One option would be to give it a type like `{a} (Literal 6 a, Literal 7 a, Literal 8 a) => [3]a`, but that scales very poorly for long sequences.

**Comment by LeventErkok** [comment]
Good point. I remember this being a big deal in Cryptol-1 as the type-checker would simply go out-to-lunch with "gold test-vectors" that people copied and pasted from PDF files.

The other option is to not allow polymorphism at all when you have literals, and ask the user to annotate with a mono-type. That sounds rather limiting, but in practice I don't think there's much use case for a polymorphic list of literals. At least I can't think of any use cases where you'd use it at two different types.

**Comment by brianhuffman** [comment]
I think there's rather a lot of use for polymorphic functions that use a literal like `0` or `1` or `2` somewhere in the body.

**Comment by LeventErkok** [comment]
I meant a list consisting only of literals. Literals by themselves make a lot of sense polymorphically, but a list of them together being used at two different types? I'm not sure if I've ever seen a good use case for that.

**Comment by brianhuffman** [comment]
I've definitely written polymorphic cryptol functions with expressions like `zero # [1]` before. I think any special typing rule for lists of literals is going to seem really arbitrary, inconsistent, and restrictive to somebody, no matter how you define it.

**Comment by brianhuffman** [comment]
Although I suppose you could just require people to use the "magic back tick" if they want a polymorphic literal, and make all other literals monomorphic.

**Comment by LeventErkok** [comment]
Agreed. No matter how you do this, it won't please everyone. But the mono-literal-list requirement is the least surprising option I think: That is, it might be annoying to annotate, but at least you'll be sure what you write is exactly what you get.
